pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -106,182 +180,237 @@ func (r *reconciler) Reconcile(rule *CompletedRule) error {
-			addedFrom = append(addedFrom, ipAddressToOFAddress(a.IP))
-		}
-		for _, a := range lastRealized.FromAddresses.Difference(newRule.FromAddresses) {
-			deletedFrom = append(deletedFrom, ipAddressToOFAddress(a.IP))
-		}
-		newOFPorts = r.podsToOFPorts(newRule.Pods)
-		for p := range newOFPorts.Difference(lastRealized.podOFPorts) {
-			addedTo = append(addedTo, openflow.NewOFPortAddress(p))
-		}
-		for p := range lastRealized.podOFPorts.Difference(newOFPorts) {
-			deletedTo = append(deletedTo, openflow.NewOFPortAddress(p))
+		from1 := podsToOFAddresses(newRule.FromAddresses)
+		from2, exceptFrom := ipBlocksToOFAddresses(newRule.From.IPBlocks)
+		addedFrom := podsToOFAddresses(newRule.FromAddresses.Difference(lastRealized.FromAddresses))
+		deletedFrom := podsToOFAddresses(lastRealized.FromAddresses.Difference(newRule.FromAddresses))
+
+		podsByServicesMap, servicesMap := groupPodsByServices(newRule.Services, newRule.Pods)
+		for svcHash, pods := range podsByServicesMap {
+			newOFPorts := r.getPodOFPorts(pods)
+			ofID, exists := lastRealized.ofIDs[svcHash]
+			// Install a new Openflow rule if this group doesn't exist, otherwise do incremental update.
+			if !exists {
+				ofRule := &types.PolicyRule{
+					Direction:  networkingv1.PolicyTypeIngress,
+					From:       append(from1, from2...),
+					ExceptFrom: exceptFrom,
+					To:         ofPortsToOFAddresses(newOFPorts),
+					Service:    servicesToNetworkPolicyPort(servicesMap[svcHash]),
+				}
+				ofID, err := r.installOFRule(ofRule)
+				if err != nil {
+					return err
+				}
+				lastRealized.ofIDs[svcHash] = ofID
+			} else {
+				addedTo := ofPortsToOFAddresses(newOFPorts.Difference(lastRealized.podOFPorts[svcHash]))
+				deletedTo := ofPortsToOFAddresses(lastRealized.podOFPorts[svcHash].Difference(newOFPorts))
+				if err := r.updateOFRule(ofID, addedFrom, addedTo, deletedFrom, deletedTo); err != nil {
+					return err
+				}
+				// Delete valid servicesHash from staleOFIDs.
+				delete(staleOFIDs, svcHash)
+			}
+			lastRealized.podOFPorts[svcHash] = newOFPorts
 		}
 	} else {
-		newIPs = r.podsToIPs(newRule.Pods)
-		for ip := range newIPs.Difference(lastRealized.podIPs) {
-			addedFrom = append(addedTo, ipStringToOFAddress(ip))
-		}
-		for ip := range lastRealized.podIPs.Difference(newIPs) {
-			deletedFrom = append(addedTo, ipStringToOFAddress(ip))
+		newIPs := r.getPodIPs(newRule.Pods)
+		from := ipsToOFAddresses(newIPs)
+		addedFrom := ipsToOFAddresses(newIPs.Difference(lastRealized.podIPs))
+		deletedFrom := ipsToOFAddresses(lastRealized.podIPs.Difference(newIPs))
+
+		podsByServicesMap, servicesMap := groupPodsByServices(newRule.Services, newRule.ToAddresses)
+		// If IPBlocks is defined, ensure its corresponding servicesHash is present in podsByServicesMap,
+		// so its Openflow rule can be updated too.
+		if len(newRule.To.IPBlocks) > 0 {
+			// IPBlocks can't resolve named port, filter out them.
```

Comment:
nit: filter them out.

Reply:
Done.

Reply:
Found a bug if filtering named ports out here that IPBlocks will be treated as allowing all services, removed the filtering process so it's same as Pods that cannot resolve any named ports.